### ISSUE TITLE
some interactions identified by Liana are not present in the cust_omni tibble obtained after running generate_omni

### ISSUE THREAD

**Issue by paulyashna** [saezlab/liana#29]
Hi Daniel and authors,

1) I have generated a customized omnipath resource to be used as input with liana_wrap as follows:

cust_omni <- generate_omni(loc_consensus_percentile = 30,
                           consensus_percentile = NULL,
                           transmitter_topology = c('secreted',
                                                    'plasma_membrane_transmembrane',
                                                    'plasma_membrane_peripheral'),
                           receiver_topology = c('plasma_membrane_transmembrane',
                                                 'plasma_membrane_peripheral'),
                           min_curation_effort = 1,
                           ligrecextra = FALSE,
                           remove_complexes = FALSE,
                           simplify = FALSE)

As a result this outputs a tibble of 6403 interactions. After running Liana though with this resource, I get some interactions that do not overlap with this custom generated omnipath resource. Would you have an idea what's going on. Ideally, Liana should not result into interactions not present in the resource. right?

2) Secondly, I am not able to run 'sca' as a part of liana_wrap or alone on more than 70k cells. Is there a parameter that couldbe modulated for this? At least the default parameters of sca do not allow this in a straight forward way.

How I run Liana: 

liana_test6methods <- liana_wrap(seurat_object = sobj,
                         resource='custom',
                         method = c('logfc','natmi','connectome','cellphonedb','sca','cellchat'),
                         external_resource = cust_omni,
                         parallelize = TRUE,
                         liana_call.params=list(complex_policy = "mean"))

Appreciate your help on these doubts.
Thanks,
Yashna

**Comment by dbdimitrov** [comment]
Hi @paulyashna,

Thanks for the issue. 

1. Actually, in one of the recent updates of LIANA, I filtered a few more interactions (namely such which have ambiguous receptors from liana's default resource). So, if you run the resource above, you should indeed expect to see some differences. 

I am planning a subsequent improvement iteration of LIANA's resource, namely one that focuses solely on interactions curated in the context of CCC resources (this is based on some work that we did as part of our revisions). You can see some information in one of the functions that we added in OmnipathR:
`OmnipathR::curated_ligand_receptor_interactions()` or https://github.com/saezlab/OmnipathR/blob/43dcfe523dfde6c03e381d800221614c59b38930/R/intercell_curated.R#L109

2. Is the error by any chance related to 'allocated memory'? If you could please share the error, I'd be happy to look more into this.

Best wishes,
Daniel

**Comment by paulyashna** [comment]
Hi Daniel,

Thanks for the quick reply and pointing me towards the improved function. I will have a look at it.

Regarding the 'sca', there is no error actually. In fact it does create a resulting liana_result$sca tibble. It's just empty. And so there is no error even in the next aggregate_liana step. However, on the same dataset at <80k cells and using the above mentioned custom_omni resource  it produces a normal populated 'sca' tibble.

3) Sorry, I just had another observation in last 3-4 hours during my downstream analysis which I would like to share. I have found some duplicated interactions in the Liana result and these duplicated interactions have the same interaction score. This was also not happening before, and I am not sure what should be causing this.

[LianaDuplicates.xlsx](https://github.com/saezlab/liana/files/8016924/LianaDuplicates.xlsx)

Please find the attached excel sheet with a small output from my data with the duplicates

**Comment by dbdimitrov** [comment]
Hi Yashna,

I expect liana to behave the way that you suggest when it aggregates the results, however I'm really not sure how it could work with one resource, but not the other. Also, it's odd that it is somehow specific to SCA alone. I assume it is related to the `global_mean` that is used in to calculate all interactions in SCA, and possibly for some reason when you use the inbuilt resource it results in the `global_mean` being a `NA`. 

One way to double check if this is the case would be to:
1. Filter your Seurat/SCE object to `union(cust_omni$ligand, cust_omni$receptor)`
2. Then sum the logcounts/data from your object.

I just wonder whether this is related to the size of the object, or if there is a NA value in your matrix. 
Knowing this, I could make safety checks to fix this error. Regardless, I will try to work on preventing an error from either case, if you can't double check yourself :)

Regarding the duplicates, it might be worth it to double check ligand_complex and receptor_complex. ITGB8 is possibly part of a complex and the way that LIANA works is that it returns the symbol with the lowest score in ligand, while ligand_complex stores the whole information

**Comment by dbdimitrov** [comment]
Hi Yashna,

1. If the sum of counts is indeed exactly 0, is it truly log normalized? If it is then this is the reason why you see NAs from SCA.
2. Technically, the way that LIANA works is that I "decomplexify" any complexes, and then "recomplexify" each methods independently. Meaning that I calculate stats for each gene in each interaction and then combine them back into complexes.
3. Also, missing genes wouldn't affect LIANA, they simply don't map.
4. Excellent

Yes, indeed, it seems like this is likely a memory issue. 

Regarding the complexes point, it is odd that the interaction scores are exactly the same. LIANA should return the subunit with the minimum expression/method-specific score close to 0, and use a mean to calculate the scores for the interactions. I will double check this also. Nevertheless, I would suggest plotting the complex names. :)

I plan on looking into this in detail tomorrow and will hopefully update LIANA in the next couple of days.

Thanks a lot for your feedback!

Cheers,
Daniel

**Comment by paulyashna** [comment]
Hi Daniel,

I have tried to run Liana 0.0.7 with the consensus resource you have suggested. 

Good news is:  'SCA' works well now with my larger dataset of 2.7 Gb (Thanks for this!!)

Certain observations:

1. I know and see in all your tutorial's sessionInfo() that R 4.0.3 has been used. However I was using Liana 0.0.5 before I did the update to 0.0.7 today and Liana 0.0.5 was working fine with R 4.11. Liana 0.0.7 gives the following error with R 4.1.1 (You already know this most probably): 

liana_test <- liana_wrap(seuratObject)
Running LIANA with RNA as default assay
Error: BiocParallel errors
  1 remote errors, element index: 1
  7 unevaluated and other errors
  first remote error: invalid length of row names

2. When I am able to run it on R 4.0.3 (!actually R 4.0.4 - this is the only env I have :P !), cellchat does not work as a part of liana_wrap. I think you know this as cellchat results do no show up here https://saezlab.github.io/liana/articles/liana_tutorial.html in  'Aggregate and Obiain Consensus Ranks'. I would however like to include cellchat results as well. When running cellchat separately, for my data it gives a blank error, for Liana testdata it gives: 


liana_testFromLiana2 <- liana_wrap(testdata,method = 'cellchat')
Running LIANA with RNA as default assay
Now Running: Cellchat
Create a CellChat object from a data matrix
Set cell identities for the new CellChat object
Error: Cell labels cannot contain `0`!  (**for my data this line is blank after Error:** )
Warning message:
In .filter_sce(sce) :
  4368 genes and/or 0 cells were removed as they had no counts!

3. I would like to test the consensus resource as you suggested but was not able to figure out - for the manually curated CCCs how can I include only 'transmembrane proteins' for both transmitter and receiver and thresholds for number of curation references etc. I tried doing so by combining the get_curated_omni() and generate_omni() functions. I actually get the same number of interactions. Is it safe to assume that this curated CCC consensus resource already includes the filters of 'OmniPath resource as it is in the Biorxiv'. 

I thank you a lot for all your guidance on these problems!

Best Regards,
Yashna

**Comment by dbdimitrov** [comment]
Hi Yashna,

1. This issue is related to incompatibilities between certain BioConductor packages on 4.1. See here for a fix:
https://github.com/saezlab/liana/issues/25

2. LIANA will return blank for the error unless you check the returned list by name. This error is thrown by CellChat it seems on purpose, I've seen it myself. A way around it is to simply rename your clusters so that neither is '0'.

3. Yes, the CCC consensus includes the same filters as in the Biorxiv + I also do a number of additional quality control steps which I thought are out of the scope for the average user and have hence not included it in the tutorial.
(Technically, they are very similar in size, but they should be a bit different also).

I documented them in detail and the function is available here:
https://github.com/saezlab/liana/blob/976659316aa3574993822eac51bd1b2633bdd0de/R/get_omnipath.R#L659

a) You could reproduce it by changing it to keep trans-membrane proteins for both the transmitter and receiver.

b) You could obtain all trans-membrane-only interactions from the `generate_omni` and filter them down to those which in the curated resource.

**Comment by paulyashna** [comment]
Hi Daniel,

I now updated all packages on R v4.1.1, and Liana works smoothly. Thanks for pointing to the earlier issue.

Regarding cellchat: I actually have cluster annotations and not numbers that the seurat Idents are assigned to. Sorry I failed to mention this in my last comment. With cellchat I am experiencing a constant isssue on R v4.0.3 and R v4.1.1 which is:

> liana_test2 <- liana_wrap(Banner.big, resource='Consensus', method = 'cellchat')
Running LIANA with RNA as default assay
Now Running: Cellchat
Create a CellChat object from a data matrix
Set cell identities for the new CellChat object
The cell groups used for CellChat analysis are  A B C D E F G H
Registered S3 method overwritten by 'cli':
  method     from         
  print.boxx spatstat.geom
**Error: Not a graph object**
Warning messages:
1: In .filter_sce(sce) :
  3955 genes and/or 0 cells were removed as they had no counts!
2: Predicate functions must be wrapped in `where()`

With this working I could really then test if the complexes and interactions in the consensus resource are reported as expected in the Liana output.

Thanks for your time on this. I really appreciate it!

Best Regards,
Yashna
